server/src/main/java/org/candlepin/controller/Entitler.java

```diff
@@ -205,6 +248,68 @@ public void adjustEntitlementQuantity(Consumer consumer, Entitlement ent,
         }
     }
 
+    private long getPoolInterval(Product prod) {
+        long interval = maxDevLifeDays;
+        String prodThenString = prod.getAttributeValue("expired_after");
+        if (prodThenString != null && Long.parseLong(prodThenString) < maxDevLifeDays) {
+            interval = Long.parseLong(prodThenString);
+        }
+        return 1000 * 60 * 60 * 24 * interval;
+    }
+
+    private Pool getDevPool(Consumer consumer, String sku) {
+        PoolFilterBuilder poolFilters = new PoolFilterBuilder();
+        poolFilters.addAttributeFilter(Pool.REQUIRES_CONSUMER_ATTRIBUTE, consumer.getUuid());
+        Page<List<Pool>> poolsPage = poolManager.listAvailableEntitlementPools(consumer, null,
+                consumer.getOwner(), null, null, true, true, poolFilters, null);
+        if (poolsPage != null &&
+            poolsPage.getPageData() != null &&
+            poolsPage.getPageData().size() == 1) {
+            return poolsPage.getPageData().get(0);
+        }
+        else {
+            return null;
+        }
+    }
+
+    protected Pool assembleDevPool(Consumer consumer, String sku) {
+        // all good. create a pool for the dev consumer
+        Set<Product> providedProducts = new HashSet<Product>();
+        Date now = new Date();
+
+        Product prod = retrieveNamedDevProduct(consumer, sku);
+        if (StringUtils.isEmpty(prod.getAttributeValue("support_level"))) {
+            // if there is no SLA, apply the default
+            prod.setAttribute("support_level", this.DEFAULT_DEV_SLA);
+            prod = productCurator.createOrUpdate(prod);
+        }
+
+        for (ConsumerInstalledProduct ip : consumer.getInstalledProducts()) {
+            providedProducts.add(retrieveNamedDevProduct(consumer, ip.getProductId()));
+        }
+
```

Comment:
Looking at lines 280 - 289:
We are doind N + 1 product lookups here via the adapter. Can we make use of the ProductAdapter.getProductsByIds(owner, ids) call so that we only hit the adapter once? It would likely create some complexity in determining what products were found, but would save on lookups.

In the error messages, we might also want to distinguish b/w not finding the SKU vs installed products.